pkg/wakaama/contrib/objects/common.c

```diff
@@ -0,0 +1,413 @@
+ * @}
+ */
+
+#include "liblwm2m.h"
+#include "lwm2m_client.h"
+#include "objects/common.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief Auxiliary structure to hold pointers that form an object link data type.
+ */
+typedef struct {
+    uint16_t *object_id;    /**< pointer to object ID */
+    uint16_t *instance_id;  /**< pointer to instance ID */
+} _objlink_t;
+
+/**
+ * @brief Get data of a given type from an object's instance.
+ *
+ * @param[in]  client_data       Pointer to the LwM2M client data.
+ * @param[in]  uri               Initialized URI structure specifying the resource to get.
+ * @param[in]  expected_type     Type of data that is expected from the specified resource.
+ * @param[out] out               Pointer where to store the data.
+ * @param[in]  out_len           Length of @p out.
+ *
+ * @return 0 on success
+ * @return <0 otherwise
+ */
+static int _get_resource_data(lwm2m_client_data_t *client_data, const lwm2m_uri_t *uri,
+                              lwm2m_data_type_t expected_type, void *out, size_t out_len)
+{
+    assert(client_data);
+    assert(uri);
+    assert(out);
+
+    int result = 0;
+    int data_num = 1;
+    lwm2m_data_t *data = NULL;
+
+    /* try to find the requested object */
+    lwm2m_object_t *object = lwm2m_get_object_by_id(client_data, uri->objectId);
+    if (!object || !object->readFunc) {
+        DEBUG("[lwm2m:get_data] could not find object with ID %d\n", uri->objectId);
+        result = -1;
+        goto out;
+    }
+
+    /* prepare a new data structure */
+    data = lwm2m_data_new(1);
+    data->id = uri->resourceId;
+
+    /* read the resource from the specified instance */
+    uint8_t res = object->readFunc(uri->instanceId, &data_num, &data, object);
+    if (res != COAP_205_CONTENT || data->type != expected_type) {
+        result = -1;
+        goto free_out;
+    }
+
+    /* process the data according to the type */
+    switch (expected_type) {
+    case LWM2M_TYPE_INTEGER:
+        *(int64_t *)out = data->value.asInteger;
+        break;
+
+    case LWM2M_TYPE_FLOAT:
+        *(double *)out = data->value.asFloat;
+        break;
+
+    case LWM2M_TYPE_BOOLEAN:
+        *(bool *)out = data->value.asBoolean;
+        break;
+
+    case LWM2M_TYPE_OBJECT_LINK:
+    {
+        _objlink_t *_out = (_objlink_t *)out;
+        *_out->object_id = data->value.asObjLink.objectId;
+        *_out->instance_id = data->value.asObjLink.objectInstanceId;
+        break;
+    }
+
+    case LWM2M_TYPE_STRING:
+    case LWM2M_TYPE_OPAQUE:
+        if (data->value.asBuffer.length > out_len) {
+            DEBUG("[lwm2m:get_data] not enough space in buffer\n");
+            result = -1;
+        }
+        else {
+            memcpy(out, data->value.asBuffer.buffer, data->value.asBuffer.length);
+        }
+        break;
+
+    default:
+        DEBUG("[lwm2m:get_data] not supported type\n");
+        result = -1;
+        break;
```

Comment:
Just in case someone else adds code right after the switch-case
```suggestion
        goto free_out;
        break;
```

Reply:
I simplified the return routine, now there is only on `out` point.